core/src/main/scala/org/apache/spark/scheduler/dynalloc/ExecutorMonitor.scala

```diff
@@ -0,0 +1,267 @@
+
+  // The following fields are an optimization to avoid having to scan all executors on every EAM
+  // schedule interval to find out which ones are timed out. They keep track of when the next
+  // executor timeout is expected to happen, and the current list of timed out executors. There's
+  // also a flag that forces the EAM task to recompute the timed out executors, in case some event
+  // arrives on the listener bus that may cause the current list of timed out executors to change.
+  //
+  // There's also per-executor state used for this purpose, so that recomputations can be triggered
+  // only when really necessary.
+  //
+  // Note that this isn't meant to, and cannot, always make the right decision about which executors
+  // are indeed timed out. For example, the EAM thread may detect a timed out executor while a new
+  // "task start" event has just been posted to the listener bus and hasn't yet been delivered to
+  // this listener. There are safeguards in other parts of the code that would prevent that executor
+  // from being removed.
+  private var nextTimeout = new AtomicLong(Long.MaxValue)
+  private var timedOutExecs = Seq.empty[String]
+  @volatile private var recomputeTimedOutExecs = false
+
+  if (idleTimeoutMs < 0) {
+    throw new SparkException(s"${DYN_ALLOCATION_EXECUTOR_IDLE_TIMEOUT.key} must be >= 0!")
+  }
+  if (storageTimeoutMs < 0) {
+    throw new SparkException(s"${DYN_ALLOCATION_CACHED_EXECUTOR_IDLE_TIMEOUT.key} must be >= 0!")
+  }
+
+  def reset(): Unit = {
+    executors.clear()
+    nextTimeout.set(Long.MaxValue)
+    timedOutExecs = Nil
+    recomputeTimedOutExecs = false
+  }
+
+  def timedOutExecutors(): Seq[String] = {
+    val now = clock.getTimeMillis()
+    if (recomputeTimedOutExecs || now >= nextTimeout.get()) {
+      recomputeTimedOutExecs = false
+
+      // Temporarily set the next timeout at Long.MaxValue. This ensures that after
+      // scanning all executors below, we know when the next timeout for non-timed out
+      // executors is (whether that update came from the scan, or from a new event
+      // arriving in a different thread).
+      nextTimeout.set(Long.MaxValue)
+
+      var newNextTimeout = Long.MaxValue
+      timedOutExecs = executors.asScala
+        .filter { case (_, exec) =>
+          val deadline = exec.timeoutAt
+          if (deadline > now) {
+            newNextTimeout = math.min(newNextTimeout, deadline)
+            exec.timedOut = false
+            false
+          } else {
+            exec.timedOut = true
+
+            // An event arriving while this scan is happening may cause the deadline for
+            // the executor to change after it was read above. Check the deadline again,
+            // and if it changed, don't consider this executor for removal yet, delaying
+            // the decision until the next check.
+            if (exec.timeoutAt > deadline) {
+              recomputeTimedOutExecs = true
```

Comment:
I'm a little confused about `recomputeTimedOutExecs`. Why do we need to eagerly  recompute timed out executors while `exec.timeoutAt` changed to be bigger ? If `exec.timeoutAt` changed to be bigger than final `newNextTimeout`, we can just follow the condition `now >= nextTimeout.get()` to decide whether we need the check. And,  instead of setting `recomputeTimedOutExecs = true`, maybe, we should just update the `newNextTimeout` comparing to changed `exec.timeoutAt`. WDYT ?

Reply:
> we should just update the newNextTimeout comparing to changed exec.timeoutAt

I think that can work and would avoid a little bit of unnecessary work. Let me do that instead.